### DISCUSSION TITLE
Support API routes outside of `pages/api`

### DISCUSSION THREAD

**Discussion by sejr** [vercel/next.js#16601] [Ideas]
Hello!

I am a huge fan of NextJS and have gone all-in, using it to build a rather complex application that has a lot of pages and API routes.

As the project has grown, I have found it difficult to navigate amongst those pages (e.g. what is actually being sent to the user) and the API routes / serverless functions. I would _really_ love if I could move the `api` folder to the root of my project, and specify some configuration telling Next to look at that path for the functions.

Has anyone else felt this to be a pain point as their project has grown?

**Comment by vinodhum** [comment]
+1 any update on this or workaround ?

**Reply by gabrielf** [reply]
I use rewrites as as a work around: https://github.com/vercel/next.js/discussions/16601#discussioncomment-14738154 which works for my use case at least.

**Comment by paulhayessilex** [comment]
+1 any update?

**Comment by akellbl4** [comment]
Hey!
I decided to use `pages/api` just like a representation of API routes and moved all logic to a different folder. Then just reexport my handlers via `pages/api` filtes.
Hope it helps.

**Reply by wrod7** [reply]
@akellbl4 can you post an example of a file in your pages/api folder?

**Reply by caio-ribeiro-pereira** [reply]
@akellbl4 please could you post some example?

**Reply by akellbl4** [reply]
file: pages/api/countries.js
```js
export * from '@/server/api/countries' // where `@/` is a alias to project root
```
file: server/api/countries.js
```js
export default function handler(req, res) {
  return res.status(200).end()
}
```

**Reply by dgurns** [reply]
Thank you, this helped. In my case it was 
```export { default } from 'api/my-function'```

**Comment by huksley** [comment]
One use case for routes outside of api/ folder is to having a links to some generated resources, for example images or calendars.
For example https://example.com/calendar/my-team-calendar.ics

**Comment by subimage** [comment]
Yes please. Coming from Rails here…API routes nested within the "pages" folder (of views) gets difficult the more complex your app becomes.

**Comment by MorganDilling** [comment]
+1 any updates on this? It would be really useful.

**Comment by alexdefaro** [comment]
+1 Any updates on this? It would be really useful at least to have the "api" folder to work from the "src" folder.

**Comment by ZoranRavic** [comment]
I hope this will finally be considered now that we have the /app folder and the /pages folder is not used for any actual pages.

**Comment by tylerforesthauser** [comment]
https://beta.nextjs.org/docs/api-reference/file-conventions/route

**Reply by ZoranRavic** [reply]
This would be great, but it does not work in the current version 13.1.6

And the link to that new https://beta.nextjs.org/docs/api-reference/file-conventions/route api is not even clickable in the sidebar:

![image](https://user-images.githubusercontent.com/11726428/219478365-ab1d24ae-8c5c-479b-89de-00f0673ef0b8.png)

Which makes me think they are not yet ready to announce it.

Although it's possible to reach https://beta.nextjs.org/docs/api-reference/file-conventions/route via their search results.

**Comment by gabrielf** [comment]
A work around is to use [rewrites](https://nextjs.org/docs/pages/api-reference/config/next-config-js/rewrites):

```typescript
// From your next.config.ts
const nextConfig: NextConfig = {
  async rewrites() {
    return [
      {
        source: '/the/path/you/want',
        destination: '/api/hello',
      },
    ]
  },
}
```

This will expose the API route implemented in `pages/api/hello.ts` on the route `/the/path/you/want`. Note that the api route is _also_ available on `/api/hello`. If it is necessary to prevent requests to `/api/hello` it is probably possible to access the original path somehow and return some error code if it starts with `/api`.
